From 578a1a221a415a7fff59eb6c3fd334a5fbc8837a Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Sun, 28 Dec 2025 17:41:14 +0800
Subject: [PATCH 22/24] [PATCH] AOSCOS: ACPI / scan: Add pwm_lookup_entry for
 PWM3 on LS7A

On LoongArch laptops using LG110 GPU for display, LS7A PWM3 is used for
backlight brightness control.  Thus we need to call pwm_add_table() so
that the loonggpu driver can find the PWM for controlling the backlight.

Signed-off-by: Xi Ruoyao <xry111@xry111.site>
Signed-off-by: darkyzhou <me@zqy.io>
---
 drivers/acpi/Makefile                        |  1 +
 drivers/acpi/internal.h                      |  6 +++
 drivers/acpi/loongarch/Makefile              |  3 ++
 drivers/acpi/loongarch/acpi-ls7a-backlight.c | 42 ++++++++++++++++++++
 drivers/acpi/scan.c                          |  1 +
 5 files changed, 53 insertions(+)
 create mode 100644 drivers/acpi/loongarch/Makefile
 create mode 100644 drivers/acpi/loongarch/acpi-ls7a-backlight.c

diff --git a/drivers/acpi/Makefile b/drivers/acpi/Makefile
index d1b0affb8..2148e6a5e 100644
--- a/drivers/acpi/Makefile
+++ b/drivers/acpi/Makefile
@@ -134,3 +134,4 @@ obj-$(CONFIG_ACPI_VIOT)		+= viot.o
 
 obj-$(CONFIG_RISCV)		+= riscv/
 obj-$(CONFIG_X86)		+= x86/
+obj-$(CONFIG_LOONGARCH)		+= loongarch/
diff --git a/drivers/acpi/internal.h b/drivers/acpi/internal.h
index 63354972a..2bf9a408e 100644
--- a/drivers/acpi/internal.h
+++ b/drivers/acpi/internal.h
@@ -76,6 +76,12 @@ void acpi_lpss_init(void);
 static inline void acpi_lpss_init(void) {}
 #endif
 
+#if IS_ENABLED(CONFIG_PWM_LOONGSON)
+void acpi_ls7a_pwm_init(void);
+#else
+static inline void acpi_ls7a_pwm_init(void) {}
+#endif
+
 void acpi_apd_init(void);
 
 acpi_status acpi_hotplug_schedule(struct acpi_device *adev, u32 src);
diff --git a/drivers/acpi/loongarch/Makefile b/drivers/acpi/loongarch/Makefile
new file mode 100644
index 000000000..25f207f84
--- /dev/null
+++ b/drivers/acpi/loongarch/Makefile
@@ -0,0 +1,3 @@
+ifdef CONFIG_PWM_LOONGSON
+obj-y += acpi-ls7a-backlight.o
+endif
diff --git a/drivers/acpi/loongarch/acpi-ls7a-backlight.c b/drivers/acpi/loongarch/acpi-ls7a-backlight.c
new file mode 100644
index 000000000..5363747cd
--- /dev/null
+++ b/drivers/acpi/loongarch/acpi-ls7a-backlight.c
@@ -0,0 +1,42 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (C) 2025 Xi Ruoyao <xry111@aosc.io>
+ *
+ * ACPI support for devices using LG110 <sarcasm>discrete</sarcasm> graphics
+ * with LS7A PWM controlling the backlight.
+ */
+
+#include <linux/acpi.h>
+#include <linux/pwm.h>
+
+#include "../internal.h"
+
+static struct pwm_lookup ls7a_pwm_lookup[] = {
+	PWM_LOOKUP_WITH_MODULE("LOON0006:03", 0, NULL, "gsgpu_backlight", 0,
+			       PWM_POLARITY_NORMAL, "pwm-loongson"),
+};
+
+static int acpi_ls7a_pwm_attach(struct acpi_device *adev,
+				 const struct acpi_device_id *id)
+{
+	if (acpi_dev_uid_match(adev, 3))
+		pwm_add_table(ls7a_pwm_lookup, ARRAY_SIZE(ls7a_pwm_lookup));
+
+	return 0;
+}
+
+static const struct acpi_device_id acpi_ls7a_pwm_ids[] = {
+	{ "LOON0006", 0 },
+	{ },
+};
+
+static struct acpi_scan_handler ls7a_pwm_handler = {
+	.ids = acpi_ls7a_pwm_ids,
+	.attach = acpi_ls7a_pwm_attach,
+};
+
+
+void __init acpi_ls7a_pwm_init(void)
+{
+	acpi_scan_add_handler(&ls7a_pwm_handler);
+}
diff --git a/drivers/acpi/scan.c b/drivers/acpi/scan.c
index ef16d58b2..c1b775fe5 100644
--- a/drivers/acpi/scan.c
+++ b/drivers/acpi/scan.c
@@ -2713,6 +2713,7 @@ void __init acpi_scan_init(void)
 	acpi_power_resources_init();
 	acpi_int340x_thermal_init();
 	acpi_init_lpit();
+	acpi_ls7a_pwm_init();
 
 	acpi_scan_add_handler(&generic_device_handler);
 
-- 
2.51.2

